spec/workflow_spec.rb

```diff
@@ -75,38 +73,156 @@
     end
   end
 
+  describe "#run_async!" do
+    it "starts the first state" do
+      workflow = make_workflow(input, {"FirstState" => {"Type" => "Wait", "Seconds" => 10, "End" => true}})
+      workflow.run_async!
+
+      expect(context.started?).to eq(true)
+    end
+
+    it "doesn't wait for the state to finish" do
+      workflow = make_workflow(input, {"FirstState" => {"Type" => "Wait", "Seconds" => 10, "End" => true}})
+      workflow.run_async!
+
+      expect(context.running?).to eq(true)
+      expect(context.ended?).to   eq(false)
+    end
+  end
+
   describe "#step" do
     it "runs a success step" do
-      input = {"input" => "value"}.freeze
-
-      workflow, ctx = make_workflow(input, {"FirstState" => {"Type" => "Succeed"}})
+      workflow = make_workflow(input, {"FirstState" => {"Type" => "Succeed"}})
 
       expect(workflow.status).to eq("pending")
       expect(workflow.end?).to eq(false)
-      expect(ctx.status).to eq("pending")
-      expect(ctx.started?).to eq(false)
-      expect(ctx.running?).to eq(false)
-      expect(ctx.ended?).to eq(false)
+      expect(context.status).to eq("pending")
+      expect(context.started?).to eq(false)
+      expect(context.running?).to eq(false)
+      expect(context.ended?).to eq(false)
 
       workflow.step
 
       expect(workflow.output).to eq(input)
       expect(workflow.status).to eq("success")
       expect(workflow.end?).to eq(true)
-      expect(ctx.output).to eq(input)
-      expect(ctx.status).to eq("success")
-      expect(ctx.started?).to eq(true)
-      expect(ctx.running?).to eq(false)
-      expect(ctx.ended?).to eq(true)
+      expect(context.output).to eq(input)
+      expect(context.status).to eq("success")
+      expect(context.started?).to eq(true)
+      expect(context.running?).to eq(false)
+      expect(context.ended?).to eq(true)
+    end
+  end
+
+  describe "#step_nonblock" do
+    context "with a workflow that hasn't started yet" do
+      it "starts the first step" do
+        workflow = make_workflow(input, {"FirstState" => {"Type" => "Wait", "Seconds" => 10, "End" => true}})
+        workflow.step_nonblock
+
+        expect(workflow.status).to eq("running")
+        expect(workflow.end?).to   eq(false)
+      end
+    end
+
+    context "with a running state" do
+      it "returns Try again" do
+        workflow = make_workflow(input, {"FirstState" => {"Type" => "Wait", "Seconds" => 10, "End" => true}})
+        expect(workflow.step_nonblock).to eq(Errno::EAGAIN)
+      end
+    end
```

Comment:
what does this buy us?

If it said it wanted 10 seconds.
And you called back in 8 seconds, and it wasn't ready, why not just have it return basically saying "call back in 2 seconds or something"?

Reply:
I wouldn't mind a method that allowed a caller to ask when to check a state again in cases where we know the ready time (wait, task+catch+retry).  I'd like to tackle this in a future enhancement, I was planning on doing something like this for the wait method that is waiting on many states and set the wakeup to the first Wait completion this would be a good way to get that time generically.